### PR TITLE
DeserializeFromJSON() method for TicketRequest object.

### PR DESCRIPTION
**Pull request by baloyan**
- Implemented the second DeserializeFromJSON() method for TicketRequest  object, it deserializes only single object not an array.
- Modified sending and receiving of ticket request message in client and cache side.

### CODE REVIEW

File: include/common/verify.h [hunk 1]
```diff
@@ -123,7 +123,7 @@ int VerifyRootFromPath(const uint8_t leaf_hash[],
 //   3. Verifying ECDSA failed.
 int VerifyECDSA(const uint8_t digest[],
                 const uint8_t pubkey[],
-                const unsigned int sig_size,
+                const long sig_size,
```
Comment (vladimir-v-diaz) [inline include/common/verify.h]: What is the reason for changing this to a long?  Should it be unsigned?

File: include/data_types/ticket_request.h [hunk 2]
```diff
@@ -75,6 +76,25 @@ struct TicketRequest {
...
+  // Method deserializes and creates TicketRequest object from JSON format,
+  // which specified in a boost::property_tree::ptree type object.
+  // JSON format should be serialized by TicketRequest's SerializeToJSON()
+  // method.
+  //
+  // Arguments:
+  // - pt: a boost::property_tree::ptree that has previously been populated with
+  //   the contents of a JSON object that defines a TicketRequest by
+  //   boost::property_tree::read_json() function.
+  // - root: the name of the key, value of which is JSON formatted TicketRequest
```
Comment (vladimir-v-diaz) [inline include/data_types/ticket_request.h]: I tried to add blank lines between the listed arguments in another pull request.  It is currently more difficult to read these comments when it is all bunched together.  I recommend you add blank lines between arguments here, and elsewhere.

File: include/data_types/ticket_request.h [hunk 3]
```diff
@@ -75,6 +76,25 @@ struct TicketRequest {
...
+  //   type object.
+  //
+  // Returns:
+  // - Deserialized TicketRequest type object.
```
Comment (vladimir-v-diaz) [inline include/data_types/ticket_request.h]: I've seen this a few times now, and tried to fix it in another pull request.  You can omit the "type" word here.  It is clear to just say, "Deserialized TicketRequest object."

You should try to fix it here and elsewhere.

File: include/dataservice/data_service_cache.h [hunk 4]
```diff
@@ -148,7 +148,7 @@ int SetDoubleEncDataMsg(CacheDataBlockResponse* datablock_response,
 // - pt: A data structure, which holds/contains TicketLotteryOneMsg object's
 //       data in a key-value form. It has previously been populated by
 //       boost::property_tree::read_json() function.
-// - datablock_response: It contains almost all needed information in order to
+// - datablock_request: It contains almost all needed information in order to
```
Comment (vladimir-v-diaz) [inline include/dataservice/data_service_cache.h]: You now name this `datablock_request`, but its type is `CacheDataBlockResponse`.  This is confusing.

File: src/common/verify.cpp [hunk 5]
```diff
@@ -367,7 +367,7 @@ int VerifyRootFromPath(const uint8_t leaf_hash[],
 //   3. Verifying ECDSA failed.
 int VerifyECDSA(const uint8_t digest[],
                 const uint8_t pubkey[],
-                const unsigned int sig_size,
+                const long sig_size,
```
Comment (vladimir-v-diaz) [inline src/common/verify.cpp]: Why long?  Should it be unsigned?

File: src/data_types/ticket_request.cpp [hunk 6]
```diff
@@ -101,9 +101,55 @@ std::string TicketRequest::SerializeToJSON() const {
 
 
 
+// Method deserializes and creates TicketRequest object from JSON format,
+// which specified in a boost::property_tree::ptree type object.
```
Comment (vladimir-v-diaz) [inline src/data_types/ticket_request.cpp]: nitpick: "which is specified in a ..."  This typo is repeated elsewhere, so I recommend you grep for the other instances.

File: src/data_types/ticket_request.cpp [hunk 7]
```diff
@@ -101,9 +101,55 @@ std::string TicketRequest::SerializeToJSON() const {
...
+//   boost::property_tree::read_json() function.
+// - root: the name of the key, value of which is JSON formatted TicketRequest
+//   type object.
+//
+// Returns:
+// - Deserialized TicketRequest type object.
+TicketRequest TicketRequest::DeserializeFromJSON(const boost::property_tree::ptree &pt,
+                                                 std::string root) {
+
+  // In order to get level access if it needs.
```
Comment (vladimir-v-diaz) [inline src/data_types/ticket_request.cpp]: I don't understand this comment.

File: src/data_types/ticket_request.cpp [hunk 8]
```diff
@@ -101,9 +101,55 @@ std::string TicketRequest::SerializeToJSON() const {
...
+// Method deserializes and creates TicketRequest object from JSON format,
+// which specified in a boost::property_tree::ptree type object.
+// JSON format should be serialized by TicketRequest's SerializeToJSON()
+// method.
+//
+// Arguments:
+// - pt: a boost::property_tree::ptree that has previously been populated with
+//   the contents of a JSON object that defines a TicketRequest by
+//   boost::property_tree::read_json() function.
+// - root: the name of the key, value of which is JSON formatted TicketRequest
```
Comment (vladimir-v-diaz) [inline src/data_types/ticket_request.cpp]: I still wonder about what this name is used for?  Maybe an example of how one is expected to use a different root key might help.

File: src/data_types/tickets_bundle.cpp [hunk 9]
```diff
@@ -71,15 +71,15 @@ TicketsBundle::~TicketsBundle() {
 
 
 // Method serializes TicketsBundle object to JSON format.
-// JSON format: "type": type,
-//              "size": size,
-//              "plain": plain message of object
+// JSON format form: "type": type,
```
Comment (vladimir-v-diaz) [inline src/data_types/tickets_bundle.cpp]: Are there missing curly braces?  I expected {"type": type, "size": size, ...}
Also, is this valid JSON?  Shouldn't it be a string?  What about encoding?  Shouldn't this be unicode, or is encoding handled elsewhere?

File: src/data_types/tickets_bundle.cpp [hunk 10]
```diff
@@ -226,17 +226,7 @@ void TicketsBundle::DeserializeFromJSON(const boost::property_tree::ptree &pt) {
...
-  // this will be added by the signer thread.
-  // Constant numbers show the number of built-in data types and size of arrys.
-  // See BatchSignature's data members for more details.
-  const int bsig_size  = 2 * sizeof(uint8_t) + sizeof(uint32_t) +
-                         2 * num_blocks * SHA256_HASH_SIZE +
-                         3 * SHA256_HASH_SIZE + SIGNATURE_SIZE;
-
-  // Constant numbers show the number of built-in data types and size of arrys.
-  // See TicketsBundle's data members for more details.
+  // See specs.txt
```
Comment (vladimir-v-diaz) [inline src/data_types/tickets_bundle.cpp]: Can this be a bit more specific, perhaps provide a particular subsection of specs.txt.  I'm not sure where to look in specs.txt, but I'm assuming I should find a definition of the TicketsBundle object in the specification?

File: src/dataservice/data_service_cache.cpp [hunk 11]
```diff
@@ -280,7 +280,7 @@ void HandleDataBlockRequest(HttpServer &server,
 // Process the received ticket request message, basically a datablock request.
 // This processing includes the following:
 // - Extract the client public key, the requested block index, h_inner_key, and
-//   request sequence, which a batch signature stucture.
+//   request sequence, which are a TicketRequest and BacthSignature stuctures.
```
Comment (vladimir-v-diaz) [inline src/dataservice/data_service_cache.cpp]: typo in `BacthSignature`.  I've seen this typo a few times now.

**Comment by vladimir-v-diaz** [comment]
In the pull request summary, you say, "Implemented the second DeserializeFromJSON() method ..."  What is the first?

File: include/common/verify.h (hunk 1 above)
Comment (baloyan) [inline include/common/verify.h]: Actually where is no reason, the most probably I wast testing something and left it as long, I will fix it, though in PR https://github.com/Cache-Cash/CacheCash/pull/121 everything related to data types fixed.

File: src/data_types/tickets_bundle.cpp (hunk 9 above)
Comment (baloyan) [inline src/data_types/tickets_bundle.cpp]: I think better to show like this `{"type": type, "size": size, ...}`, and it is a valid JSON string.
In PR https://github.com/Cache-Cash/CacheCash/pull/119 I construct JSON string using boost property tree object, which does what you mentioned here.

**Comment by baloyan** [comment]
The first one is here, which is already merged https://github.com/Cache-Cash/CacheCash/pull/85

File: src/common/verify.cpp (hunk 5 above)
Comment (baloyan) [reply src/common/verify.cpp]: The reason is we use `sig_size` variable (as a last argument) only when we call `d2i_ECDSA_SIG()` function and `d2i_ECDSA_SIG()` function's last argument type is long.
See: https://www.openssl.org/docs/man1.0.2/crypto/d2i_ECDSA_SIG.html

File: include/data_types/ticket_request.h (hunk 2 above)
Comment (baloyan) [reply include/data_types/ticket_request.h]: I might be do this in a separate PR. For example, to me it is not readable if it has blank line, but it is only my personal view not a general.

**Comment by vladimir-v-diaz** [comment]
> The first one is here #85, which was already merged.
I recommend that you use explicit statements in your pull request summaries and issues.  For instance, instead of saying "implemented the second ..." you could instead link to the second function to avoid ambiguity.

I find that explicit statements help with readability, and comments that use words like  "this", "that", etc. lead to confusion.

File: src/common/verify.cpp (hunk 5 above)
Comment (vladimir-v-diaz) [reply src/common/verify.cpp]: I see.  Maybe you should explain that in a comment.  Reading that line on its own, it's unclear why a sig needs negative values.

Comment (baloyan) [reply src/common/verify.cpp]: Added description in a comment.

File: include/dataservice/data_service_cache.h (hunk 4 above)
Comment (baloyan) [reply include/dataservice/data_service_cache.h]: Oops!, it has to be `datablock_response`

File: src/data_types/tickets_bundle.cpp (hunk 10 above)
Comment (baloyan) [reply src/data_types/tickets_bundle.cpp]: In this commit I specified it in details 
https://github.com/baloyan/CacheCash/commit/a741f44ab190c3b7bb80f3ebf8898c0e08bb1862

**Pull request merged**